File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/model/IDODrinkWaterRemindModel.kt
package com.idosmart.model

import com.google.gson.GsonBuilder
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import java.lang.reflect.Type

open class IDODrinkWaterRemindModel(
    onOff: Int,
    startHour: Int,
    startMinute: Int,
    endHour: Int,
    endMinute: Int,
    repeats: HashSet<IDOWeek>,
    interval: Int,
    notifyFlag: Int?,
    doNotDisturbOnOff: Int?,
    noDisturbStartHour: Int?,
    noDisturbStartMinute: Int?,
    noDisturbEndHour: Int?,
    noDisturbEndMinute: Int?
) : IDOBaseAdapterModel<IDODrinkWaterRemindModel> {
    // Switch 0: Off 1: On
    var onOff = onOff

    // Start time Hour
    var startHour = startHour

    // Start time Minute
    var startMinute = startMinute

    // End time Hour
    var endHour = endHour

    // End time Minute
    var endMinute = endMinute

    // Repeat
    var repeats: HashSet<IDOWeek> = repeats

    // Reminder interval in minutes
    var interval = interval

    // Notification type
    // 0: Invalid
    // 1: Allow notification
    // 2: Silent notification
    // 3: Notification off
    // Need to open firmware table support setDrinkWaterAddNotifyFlag
    var notifyFlag = notifyFlag

    // Do not disturb switch
    // 00: Off
    // 01: On
    // Need to open firmware table support setNoReminderOnDrinkReminder
    var doNotDisturbOnOff = doNotDisturbOnOff

    // Do not disturb start time Hour
    // Need to open firmware table support setNoReminderOnDrinkReminder
    var noDisturbStartHour = noDisturbStartHour

    // Do not disturb start time Minute
    // Need to open firmware table support setNoReminderOnDrinkReminder
    var noDisturbStartMinute = noDisturbStartMinute

    // Do not disturb end time Hour
    // Need to open firmware table support setNoReminderOnDrinkReminder
    var noDisturbEndHour = noDisturbEndHour

    // Do not disturb end time Minute
    // Need to open firmware table support setNoReminderOnDrinkReminder
    var noDisturbEndMinute = noDisturbEndMinute


    override fun getDeSerializer(): IDOBaseModelDeSerializer<IDODrinkWaterRemindModel>? {
        return IDODrinkWaterRemindModelSerializer()
    }

    override fun toString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }


}

internal class IDODrinkWaterRemindModelSerializer :
    IDOBaseModelDeSerializer<IDODrinkWaterRemindModel> {
    override fun serialize(
        src: IDODrinkWaterRemindModel,
        typeOfSrc: Type?,
        context: JsonSerializationContext?
    ): JsonElement {
        val jsonObject = JsonObject()
        jsonObject.addProperty("on_off", src.onOff)
        jsonObject.addProperty("start_hour", src.startHour)
        jsonObject.addProperty("start_minute", src.startMinute)
        jsonObject.addProperty("end_hour", src.endHour)
        jsonObject.addProperty("end_minute", src.endMinute)
        jsonObject.addProperty("interval", src.interval)
        jsonObject.addProperty("notify_flag", src.notifyFlag)
        jsonObject.addProperty("do_not_disturb_on_off", src.doNotDisturbOnOff)
        jsonObject.addProperty("no_disturb_start_hour", src.noDisturbStartHour)
        jsonObject.addProperty("no_disturb_start_minute", src.noDisturbStartMinute)
        jsonObject.addProperty("no_disturb_end_hour", src.noDisturbEndHour)
        jsonObject.addProperty("no_disturb_end_minute", src.noDisturbEndMinute)

        var num = 0
        src.repeats.forEach { week ->
            num = num or (1 shl week.rawValue + 1)
        }
        jsonObject.addProperty("repeat", num)
        return jsonObject
    }

    override fun deserialize(
        json: JsonElement?,
        typeOfT: Type?,
        context: JsonDeserializationContext?
    ): IDODrinkWaterRemindModel? {
        json?.let {
            val jsonObject = it.asJsonObject
            val onOff = jsonObject.get("on_off").asInt
            val startHour = jsonObject.get("start_hour").asInt
            val startMinute = jsonObject.get("start_minute").asInt
            val endHour = jsonObject.get("end_hour").asInt
            val endMinute = jsonObject.get("end_minute").asInt
            val interval = jsonObject.get("interval").asInt
            val notifyFlag = jsonObject.get("notify_flag").asInt
            val doNotDisturbOnOff = jsonObject.get("do_not_disturb_on_off").asInt
            val noDisturbStartHour = jsonObject.get("no_disturb_start_hour").asInt
            val noDisturbStartMinute = jsonObject.get("no_disturb_start_minute").asInt
            val noDisturbEndHour = jsonObject.get("no_disturb_end_hour").asInt
            val noDisturbEndMinute = jsonObject.get("no_disturb_end_minute").asInt

            val repeat = jsonObject.get("repeat").asInt
            val items = HashSet<IDOWeek>()
            for (i in 0..6) {
                if (repeat and (1 shl (i + 1)) != 0) {
                    items.add(IDOWeek.fromRawValue(i))
                }
            }

            return IDODrinkWaterRemindModel(
                onOff,
                startHour,
                startMinute,
                endHour,
                endMinute,
                items,
                interval,
                notifyFlag,
                doNotDisturbOnOff,
                noDisturbStartHour,
                noDisturbStartMinute,
                noDisturbEndHour,
                noDisturbEndMinute
            )
        }
        return null

    }
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/Model/IDODrinkWaterRemindModel.swift
//
//  IDODrinkWaterRemindModel.swift
//  protocol_channel
//
//  Created by hc on 2024/1/24.
//

import Foundation

// MARK: - IDODrinkWaterRemindModel
public class IDODrinkWaterRemindModel: NSObject, IDOBaseModel {
    /// Switch 0: Off 1: On
    public var onOff: Int
    /// Start time Hour
    public var startHour: Int
    /// Start time Minute
    public var startMinute: Int
    /// End time Hour
    public var endHour: Int
    /// End time Minute
    public var endMinute: Int
    /// Repeat
    public var repeats: Set<IDOWeek>
    /// Reminder interval in minutes
    public var interval: Int
    /// Notification type
    /// 0: Invalid
    /// 1: Allow notification
    /// 2: Silent notification
    /// 3: Notification off
    /// Need to open firmware table support setDrinkWaterAddNotifyFlag
    public var notifyFlag: Int
    /// Do not disturb switch
    /// 00: Off
    /// 01: On
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var doNotDisturbOnOff: Int
    /// Do not disturb start time Hour
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var noDisturbStartHour: Int
    /// Do not disturb start time Minute
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var noDisturbStartMinute: Int
    /// Do not disturb end time Hour
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var noDisturbEndHour: Int
    /// Do not disturb end time Minute
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var noDisturbEndMinute: Int

    enum CodingKeys: String, CodingKey {
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case repeats = "repeat"
        case interval = "interval"
        case notifyFlag = "notify_flag"
        case doNotDisturbOnOff = "do_not_disturb_on_off"
        case noDisturbStartHour = "no_disturb_start_hour"
        case noDisturbStartMinute = "no_disturb_start_minute"
        case noDisturbEndHour = "no_disturb_end_hour"
        case noDisturbEndMinute = "no_disturb_end_minute"
    }

    public init(onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, repeats: Set<IDOWeek>, interval: Int, notifyFlag: Int = 0, doNotDisturbOnOff: Int = 0, noDisturbStartHour: Int = 0, noDisturbStartMinute: Int = 0, noDisturbEndHour: Int = 0, noDisturbEndMinute: Int = 0) {
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.repeats = repeats
        self.interval = interval
        self.notifyFlag = notifyFlag
        self.doNotDisturbOnOff = doNotDisturbOnOff
        self.noDisturbStartHour = noDisturbStartHour
        self.noDisturbStartMinute = noDisturbStartMinute
        self.noDisturbEndHour = noDisturbEndHour
        self.noDisturbEndMinute = noDisturbEndMinute
    }
    
    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        notifyFlag = try container.decode(Int.self, forKey: .notifyFlag)
        interval = try container.decode(Int.self, forKey: .interval)
        doNotDisturbOnOff = try container.decode(Int.self, forKey: .doNotDisturbOnOff)
        noDisturbStartHour = try container.decode(Int.self, forKey: .noDisturbStartHour)
        noDisturbStartMinute = try container.decode(Int.self, forKey: .noDisturbStartMinute)
        noDisturbEndHour = try container.decode(Int.self, forKey: .noDisturbEndHour)
        noDisturbEndMinute = try container.decode(Int.self, forKey: .noDisturbEndMinute)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = Set<IDOWeek>()
            for i in 0...6 {
                if value & (1 << (i + 1)) != 0 {
                    items.insert(IDOWeek(rawValue: i)!)
                }
            }
            repeats = items
        } else {
            repeats = Set<IDOWeek>()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(interval, forKey: .interval)
        try container.encode(doNotDisturbOnOff, forKey: .doNotDisturbOnOff)
        try container.encode(noDisturbStartHour, forKey: .noDisturbStartHour)
        try container.encode(noDisturbStartMinute, forKey: .noDisturbStartMinute)
        try container.encode(noDisturbEndHour, forKey: .noDisturbEndHour)
        try container.encode(noDisturbEndMinute, forKey: .noDisturbEndMinute)
        try container.encode(notifyFlag, forKey: .notifyFlag)

        var value = 0
        for item in repeats {
            value |= (1 << (item.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }

    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}


// MARK: - IDODrinkWaterRemindModelObjc
@objcMembers
public class IDODrinkWaterRemindModelObjc: NSObject, IDOBaseModel {
    /// Switch 0: Off 1: On
    public var onOff: Int
    /// Start time Hour
    public var startHour: Int
    /// Start time Minute
    public var startMinute: Int
    /// End time Hour
    public var endHour: Int
    /// End time Minute
    public var endMinute: Int
    /// Repeat
    public var repeats: [IDOWeekObjc]
    /// Reminder interval in minutes
    public var interval: Int
    /// Notification type
    /// 0: Invalid
    /// 1: Allow notification
    /// 2: Silent notification
    /// 3: Notification off
    /// Need to open firmware table support setDrinkWaterAddNotifyFlag
    public var notifyFlag: Int
    /// Do not disturb switch
    /// 00: Off
    /// 01: On
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var doNotDisturbOnOff: Int
    /// Do not disturb start time Hour
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var noDisturbStartHour: Int
    /// Do not disturb start time Minute
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var noDisturbStartMinute: Int
    /// Do not disturb end time Hour
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var noDisturbEndHour: Int
    /// Do not disturb end time Minute
    /// Need to open firmware table support setNoReminderOnDrinkReminder
    public var noDisturbEndMinute: Int

    enum CodingKeys: String, CodingKey {
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case repeats = "repeat"
        case interval = "interval"
        case notifyFlag = "notify_flag"
        case doNotDisturbOnOff = "do_not_disturb_on_off"
        case noDisturbStartHour = "no_disturb_start_hour"
        case noDisturbStartMinute = "no_disturb_start_minute"
        case noDisturbEndHour = "no_disturb_end_hour"
        case noDisturbEndMinute = "no_disturb_end_minute"
    }

    public init(onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, repeats: [IDOWeekObjc], interval: Int, notifyFlag: Int = 0, doNotDisturbOnOff: Int = 0, noDisturbStartHour: Int = 0, noDisturbStartMinute: Int = 0, noDisturbEndHour: Int = 0, noDisturbEndMinute: Int = 0) {
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.repeats = repeats
        self.interval = interval
        self.notifyFlag = notifyFlag
        self.doNotDisturbOnOff = doNotDisturbOnOff
        self.noDisturbStartHour = noDisturbStartHour
        self.noDisturbStartMinute = noDisturbStartMinute
        self.noDisturbEndHour = noDisturbEndHour
        self.noDisturbEndMinute = noDisturbEndMinute
    }
    
    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        notifyFlag = try container.decode(Int.self, forKey: .notifyFlag)
        interval = try container.decode(Int.self, forKey: .interval)
        doNotDisturbOnOff = try container.decode(Int.self, forKey: .doNotDisturbOnOff)
        noDisturbStartHour = try container.decode(Int.self, forKey: .noDisturbStartHour)
        noDisturbStartMinute = try container.decode(Int.self, forKey: .noDisturbStartMinute)
        noDisturbEndHour = try container.decode(Int.self, forKey: .noDisturbEndHour)
        noDisturbEndMinute = try container.decode(Int.self, forKey: .noDisturbEndMinute)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = [IDOWeekObjc]()
            for i in 0...6 {
                if value & (1 << (i + 1)) != 0 {
                    items.append(IDOWeekObjc(week: IDOWeek(rawValue: i)!))
                }
            }
            repeats = items
        } else {
            repeats = [IDOWeekObjc]()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(interval, forKey: .interval)
        try container.encode(doNotDisturbOnOff, forKey: .doNotDisturbOnOff)
        try container.encode(noDisturbStartHour, forKey: .noDisturbStartHour)
        try container.encode(noDisturbStartMinute, forKey: .noDisturbStartMinute)
        try container.encode(noDisturbEndHour, forKey: .noDisturbEndHour)
        try container.encode(noDisturbEndMinute, forKey: .noDisturbEndMinute)

        var value = 0
        for item in repeats {
            value |= (1 << (item.week.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }

    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}
